src/plan/aggregate.rs

```diff
@@ -43,94 +52,201 @@ pub struct Aggregate<P: Implementable> {
-                                if min > val { min = val; }
+                                if min > val {
+                                    min = val;
+                                }
                             }
                             // @TODO could preserve multiplicity of smallest value here
                             output.push((*min, 1));
                         })
-                        .map(|(key, min)| key.iter().cloned()
-                             .chain([Value::Number(min)].iter().cloned())
-                             .collect())
+                        .map(|(key, min)| {
+                            let mut v = key.clone();
+                            v.push(Value::Number(min as i64));
+                            v
+                        }),
                 }
-            },
+            }
             AggregationFn::MAX => {
                 SimpleRelation {
                     symbols: self.variables.to_vec(),
                     tuples: tuples
-                        .map(|(ref key, ref tuple)| {
-                            let v = match tuple[0] {
-                                Value::Number(num) => num,
-                                _ => panic!("MAX can only be applied on type Number.")
-                            };
-                            (key.clone(), v)
-                        })
+                        .map(prepare_unary)
                         .group(|_key, vals, output| {
                             let mut max = vals[0].0;
                             for &(val, _) in vals.iter() {
-                                if max < val { max = val; }
+                                if max < val {
+                                    max = val;
+                                }
                             }
                             // @TODO could preserve multiplicity of largest value here
                             output.push((*max, 1));
                         })
-                        .map(|(key, max)| key.iter().cloned()
-                             .chain([Value::Number(max)].iter().cloned())
-                             .collect())
-                }
-            },
-            AggregationFn::COUNT => {
-                SimpleRelation {
-                    symbols: self.variables.to_vec(),
-                    tuples: tuples,
-                        .map(|(ref _key, ref _tuple)| ())
-                        .count()
-                        .map(|(_, count)| vec![Value::Number(count as i64)])
+                        .map(|(key, max)| {
+                            let mut v = key.clone();
+                            v.push(Value::Number(max as i64));
+                            v
+                        }),
                 }
+            }
+            AggregationFn::COUNT => SimpleRelation {
+                symbols: self.variables.to_vec(),
+                tuples: tuples
+                    .consolidate()
```

Comment:
For what it's worth, a `consolidate` before a `group` is strictly redundant. The `group` operation will perform the consolidation as part of arranging its input, using the same machinery (internally, the `arrange` operator), and in this case `consolidate` just does extra work.

This is true in a few other spots here, e.g. just before the `distinct()`, and perhaps just before the `map(prepare_unary)` calls. The latter are not technically redundant (perhaps `prepare_unary` is very expensive; consolidation would make sense), but you could almost surely leave them out and it would be fine.

Reply:
Nice!
Got it. I presumably got a bit over my head trying to ensure set semantics :)

Reply:
Consolidate is a bit of a weird operator. It is a logical "no op", in that it doesn't (shouldn't) change the collection it is applied to. Instead, it ensures that each `(data, time)` pair results in at most one `(data, time, diff)` triple, and only when `diff` is non-zero.

It's most commonly used i. as part of iteration, to make sure that you stop iterating when things change, and ii. just before an `inspect` so that you don't print out things that have logically cancelled with each other.

To ensure set semantics, `distinct` is what you would want. However, it is surprisingly expensive (it's ok, but it is surprising that it is not cheap). I'd recommend doing as little of it as you can, for example on the second argument to antijoin and perhaps when return results to users.

Reply:
Ah okay, I thought that as we are not using the `count` in the `group()` by returning `output.push((val, 1))` a single `consolidate()` beforehand would move all duplicate `(data, time)` to the `diff`. So we would need no `distinct()`
But as you mentioned before, `group()` does already take care of that